### PR TITLE
Add element

### PR DESCRIPTION
**Pull request by sofiazcoaga**
This PR includes the necessary changes to admit the possibility to add an element to an existing tree. 
New relevant functions: 
- `add_element()` : allows to include a new element in an existing merkle tree 
- `extend_tree()`: handles the case when the tree's leaves are all full of real data and the tree needs to be extended in order to add a new element. 
- `modify_existing_leaf_and_update_parents()`: handles the case where at least one of the leaves of the tree is filled with default data and the new element replaces it and updates the tree all the way up (recalculating parents' hashes). 
- `create_merkle_tree_from_leaves()`: allows the creation of a merkle tree from its leaves (hashes of the data elements) rather than the elements themselves. This function becomes useful when extending a merkle tree. 

This PR does not include a modules restructure that might occur in a later iteration.

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -1,6 +1,7 @@
+use std::vec;
+
 /// sha256::digest("") - the hash of an empty string
-const DEFAULT_HASH_OF_EMTPY: &str =
-    "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855";
+const DEFAULT_ZERO_HASH: &str = "6e340b9cffb37a989ca544e6bb780a2c78901d3fb33738768511a30617afa01d";
```
Comment (tomip01) [inline src/main.rs]: Just if you want to not be changing this you can use a crate `lazy_static` that allows you to create constant with calling a function.
`lazy_static! {
    static ref DEFAULT_ZERO_HASH: &str = sha256::digest("");
}`

Also is the comment still updated?

File: src/main.rs [hunk 2]
```diff
@@ -50,29 +51,141 @@ impl MerkleTree {
...
+        // Create the leaves level of the new subtree
+        let mut leaves = vec![sha256::digest(new_element)];
+        leaves.append(&mut vec![DEFAULT_ZERO_HASH.to_string(); fill_amount]);
+
+        // Create a new merkle tree for the new data subset
+        let mut half_merkle_tree = create_merkle_tree_from_leaves(leaves);
+
+        // Append the new merkle tree's levels to the previous tree
+        for i in 0..self.height() {
+            self.merkle_tree[i].append(&mut half_merkle_tree[i]);
```
Comment (tomip01) [inline src/main.rs]: Here you could use the function `get_level(i)`:
```suggestion
            self.merkle_tree.get_level(i).append(&mut half_merkle_tree.get_level(i));
```

File: src/main.rs [hunk 3]
```diff
@@ -50,29 +51,141 @@ impl MerkleTree {
...
 
 type Input = Vec<Vec<u8>>;
 
-pub fn create_merkle_tree_from_data(inputs: &Input) -> MerkleTree {
-    let mut merkle_tree = MerkleTree::default();
-    let data_hashes_level = create_data_hashes(inputs);
-    merkle_tree.push_level_back(data_hashes_level.clone());
-
-    let mut previous_level = data_hashes_level;
+/// Allows the creationg of a Merkle Tree from its base level (or leaves level) which are the hashes of the original
```
Comment (tomip01) [inline src/main.rs]:
```suggestion
/// Allows the creation of a Merkle Tree from its base level (or leaves level) which are the hashes of the original
```

File: src/main.rs [hunk 4]
```diff
@@ -50,29 +51,141 @@ impl MerkleTree {
...
+/// data. This function is useful to extend the Merkle Tree when adding a new element.
+fn create_merkle_tree_from_leaves(leaves: Vec<String>) -> Vec<Vec<String>> {
```
Comment (tomip01) [inline src/main.rs]: You could change the name function as it no longer returns a merkle tree

File: src/main.rs [hunk 5]
```diff
@@ -204,11 +309,56 @@ mod test {
...
+            new_amount_of_elements.next_power_of_two()
+        );
+        assert_eq!(
+            merkle_tree.leaves_amount(),
+            new_amount_of_elements.next_power_of_two()
+        );
+        assert_eq!(merkle_tree.height(), previous_height + 1);
+        assert_merkle_tree_is_consistent(merkle_tree);
+    }
+
```
Comment (tomip01) [inline src/main.rs:364]: What do you think of adding a test from adding an element to an empty merkle tree?

**Review by tomip01 — APPROVED** [review]

File: src/main.rs (hunk 1 above)
Comment (sofiazcoaga) [reply src/main.rs]: thanks for the suggestion! applied and comment updated

File: src/main.rs (hunk 3 above)
Comment (sofiazcoaga) [reply src/main.rs]: fixed

File: src/main.rs (hunk 2 above)
Comment (sofiazcoaga) [reply src/main.rs]: the thing with get_level() is that it returns an immutable reference and i need to edit the field

File: src/main.rs (hunk 4 above)
Comment (sofiazcoaga) [reply src/main.rs]: As the type `Vec<Vec<String>>` is actually the tree I made a change to mantain its semantics. What do you think of [changes committed here](https://github.com/sofiazcoaga/rusty-merkle-tree/pull/9/commits/72506aa7b9dc345c270658c1d02eef1c7c11d478) ?

File: src/main.rs (hunk 5 above)
Comment (sofiazcoaga) [reply src/main.rs:364]: actually, i think this would fail with this implementation 🤔  good observation. I will add an issue to not get it mixed up with the rest of the conversation.

File: src/main.rs (hunk 2 above)
Comment (tomip01) [reply src/main.rs]: Oh I didn't realize that! 
Just if you want to not create the `half_merkle_tree` a `mut` you can use this:
```
// Create a new merkle tree for the new data subset
let half_merkle_tree = create_merkle_tree_from_leaves(leaves);

// Append the new merkle tree's levels to the previous tree
for i in 0..self.height() {
    self.merkle_tree[i].extend_from_slice(&half_merkle_tree[i]);
}
```

File: src/main.rs (hunk 4 above)
Comment (tomip01) [reply src/main.rs]: I like that!

File: src/main.rs (hunk 2 above)
Comment (sofiazcoaga) [reply src/main.rs]: done! thanks :)
